sticker2/src/graph/mst.rs

```diff
@@ -0,0 +1,737 @@
+//! Find the maximum spanning tree using Chu Liu Edmonds
+
+use std::collections::{HashMap, HashSet};
+use std::f32;
+
+use ndarray::{ArrayView2, ArrayViewMut2, Axis};
+use ordered_float::OrderedFloat;
+
+// The implementation follows the description of:
+//
+// https://en.wikipedia.org/wiki/Edmonds%27_algorithm
+//
+// There are several differences compared to the above
+// description:
+//
+// - We want to compute the maximum spanning tree. So, we find
+//   incoming edges with maximum scores. This means that we have to
+//   change the calculation scores of incoming edges of contracted
+//   cycles. Here we follow Kübler et al., 2009, pp. 47.
+// - Since the input is a score matrix, there are no parallel edges
+//   in the input graph.
+// - Since we use (a copy of) the score matrix to store weights of
+//   incoming/outgoing contraction edges, we cannot store parallel
+//   edges. So, we only store the highest scoring parallel edge
+//   when computing the edges of the contraction. This does not change
+//   the main algorithm, since the next recursion of Chu-Lui-Edmonds
+//   would discard the lower-scoring edges anyway.
+
+/// Chu-Liu-Edmonds maximum spanning tree for dense graphs
+///
+/// This function returns the parent of each vertex in the maximum
+/// spanning tree of the `scores` square matrix, rooted at
+/// `root_vertex`.  Each row in the matrix represents outgoing edge
+/// scores of the corresponding vertex, each column incoming edge
+/// scores. Thus, `scores[(parent, child)]` should give the weight of
+/// the edge from `parent` to child.
+///
+/// Returns vertex parents the length of the returned `Vec` equals
+/// the number of rows/columns of the scores matrix.
+pub fn chu_liu_edmonds(scores: ArrayView2<f32>, root_vertex: usize) -> Vec<Option<usize>> {
+    assert_eq!(
+        scores.nrows(),
+        scores.ncols(),
+        "Score matrix must be a square matrix, has shape: ({}, {})",
+        scores.nrows(),
+        scores.ncols()
+    );
+
+    // We use this `Vec` to keep track of which vertices are 'active'.
+    // Vertices that are part of a contracted cycle become inactive.
+    let mut active_vertices = vec![true; scores.nrows()];
+
+    chu_liu_edmonds_(
+        scores.to_owned().view_mut(),
+        root_vertex,
+        &mut active_vertices,
+    )
+}
+
+fn chu_liu_edmonds_(
+    mut similarity: ArrayViewMut2<f32>,
```

Comment:
Why `similarity` and not `scores`?

Reply:
Good catch! I have changed this to `scores` throughout this function.